support custom class in torchgen schema parser

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82925

Differential Revision: [D38480514](https://our.internmc.facebook.com/intern/diff/D38480514/)

torchgen schema parser does not support parsing function schemas using custom class so far. Here is an example:
```
quantized::conv2d_relu.new(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> (Tensor)
```

This PR parse custom class name and encapsulate that into an object of CustomClassType. The only thing we need right now is just store the string class name and return that in `__str__` method.

shunting314 committed:
support custom class in torchgen schema parser



Differential Revision: [D38480514](https://our.internmc.facebook.com/intern/diff/D38480514/)

[ghstack-poisoned]

shunting314 committed:
Update on "support custom class in torchgen schema parser"


Differential Revision: [D38480514](https://our.internmc.facebook.com/intern/diff/D38480514/)

[ghstack-poisoned]

bdhirsh commented on torchgen/model.py:
```diff
@@ -1723,6 +1728,31 @@ def symint_to_int(self) -> "Type":
+    class_name: str
+
+    def __str__(self) -> str:
+        return self.class_name
```
nit: it might be clearer to store the prefix and the actual class name as separate fields on `CustomClassType`, since the `__torch__.torch.classes` bit technically isn't part of the class name.

shunting314 replied:
yep. I'll store the real class name skipping the prefix since the parsing logic assumes such a prefix anyways.

bdhirsh reviewed (approved):
small nit, but otherwise looks good to me!

This PR is enough to let you parse custom schema signatures into `NativeFunction` objects, but technically there's more you'd need to do if you want to be able to emit C++ code with custom classes using the existing codegen API's. Let me know if/when you want to talk more about that!
 technically isn't enough if you want to

shunting314 commented:
> This PR is enough to let you parse custom schema signatures into `NativeFunction` objects, but technically there's more you'd need to do if you want to be able to emit C++ code with custom classes using the existing codegen API's. Let me know if/when you want to talk more about that! technically isn't enough if you want to

Totally agree. Right now, we only need the parsing/serialization part of the Type object (so we can do the same thing for the enclosing FunctionSchema).

Will need revisit if people need full codegen functionality for custom classes. Will be glad to talk when that requirement happens :)

shunting314 committed:
Update on "support custom class in torchgen schema parser"


Differential Revision: [D38480514](https://our.internmc.facebook.com/intern/diff/D38480514/)

torchgen schema parser does not support parsing function schemas using custom class so far. Here is an example:
```
quantized::conv2d_relu.new(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> (Tensor)
```

This PR parse custom class name and encapsulate that into an object of CustomClassType. The only thing we need right now is just store the string class name and return that in `__str__` method.

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge
